### ISSUE TITLE
how to use oneSignal with hooks?

### ISSUE THREAD

**Issue by samuk190** [OneSignal/react-native-onesignal#1019]
How to access onReceived, onOpened methods if I use react hooks instead of classes?

also, I want to know how I can display notification if user received a message from someone, but don't display the message if he's talking with the user now. (same screen)

**Comment by MuriloCampos** [comment]
The only difference from the example on the docs is that you gotta use the useEffect hook instead of lifecycle methods. So you set the listeners on a useEffect (with nothing in the dependency array), and remove the listeners on the return of the hook.

**Comment by samuk190** [comment]
> The only difference from the example on the docs is that you gotta use the useEffect hook instead of lifecycle methods. So you set the listeners on a useEffect (with nothing in the dependency array), and remove the listeners on the return of the hook.

Thanks Murilo, I tried that because I did with socket io sucessfully, but with oneSignal it didnt work because I dont know how to get onReceive, onOpened..  method with Hooks Can you please show me an example?

**Comment by MuriloCampos** [comment]
just declare the callbacks as functions inside the component.

`function onOpened(openResult: any) {
    console.log('Message: ', openResult.notification.payload.body)
    console.log('Data: ', openResult.notification.payload.additionalData)
    console.log('isActive: ', openResult.notification.isAppInFocus)
    console.log('openResult: ', openResult)
  }

  function onIds(device: any) {
    console.log('Device info: ', device)
  }`

**Comment by samuk190** [comment]
> just declare the callbacks as functions inside the component.
> 
> `function onOpened(openResult: any) {
> console.log('Message: ', openResult.notification.payload.body)
> console.log('Data: ', openResult.notification.payload.additionalData)
> console.log('isActive: ', openResult.notification.isAppInFocus)
> console.log('openResult: ', openResult)
> }
> 
> function onIds(device: any) {
> console.log('Device info: ', device)
> }`

cool so in theory I can use onOpened in any screen to do my logic.. hmm if I put my oneSignal where I get access to my navigation, I can navigate user based on userData, I'll def try!

**Comment by rgomezp** [comment]
@MuriloCampos ,
Thanks for the response.

@samuk190 ,
Please let us know if the suggestion worked for you. 

We plan on improving our RN documentation regarding Hooks soon. Thank you for your patience.

**Comment by MuriloCampos** [comment]
@rgomezp NP I actually just ran into this problem a couple of weeks ago and solved it like this. @samuk190 I also forgot to add that you probably are gonna have to wrap those functions inside a useCallback hook so you can add them in the dependencies array of the useEffect.

**Comment by rubek-joshi** [comment]
I have done the following: 
```
useEffect(() => {
    OneSignal.addEventListener('received', onReceived);
    OneSignal.addEventListener('opened', onOpened);

    return () => {
      console.log('Removing Event Listeners');
      OneSignal.removeEventListener('received', onReceived);
      OneSignal.removeEventListener('opened', onOpened);
    };
  }, []);
```
Is this correct? The `console.log('Removing Event Listeners');` is never fired since I am using it in my main component so I am thinking that removing the event listener is actually unnecessary in my case. Or is it actually preventing a memory leak?

**Comment by hugoh59** [comment]
No improvement still?

**Comment by ankitgauravapporio** [comment]
hi @rgomezp can you please update the documentation , thanks

**Comment by rgomezp** [comment]
Howdy,
Apologies for the delay. We are in the process of releasing a major release across all of our SDKs. We will update this once the release has gone out. Thanks for everyone's patience.

**Comment by seftonmonk** [comment]
anyone ever solve this?

**Comment by seftonmonk** [comment]
> Howdy,
> Apologies for the delay. We are in the process of releasing a major release across all of our SDKs. We will update this once the release has gone out. Thanks for everyone's patience.

Is there any documentation for Hooks setup, I cant seem to get event listeners to fire.

**Comment by SaraChicaD** [comment]
Anyone have any examples of working OneSignal `useEffect` hooks?

**Comment by SaraChicaD** [comment]
Actually _just_ got this working, thanks to the ex from @rubek-joshi here's how I get the `onReceived` and `onOpened` to fire using `useEffect`:

```
  useEffect(() => {
    OneSignal.addEventListener('received', onReceived);
    OneSignal.addEventListener('opened', onOpened);
  }, []);

  const onOpened = openResult => {
    if (openResult) {
      console.log('Message: ', openResult.notification.payload.body);
      console.log('Data: ', openResult.notification.payload.additionalData);      
    }
  };

  const onReceived = notification => {
    console.log("Notification received: ", notification);
  };
```

> Anyone have any examples of working OneSignal `useEffect` hooks?

**Comment by devpato** [comment]
> @rgomezp NP I actually just ran into this problem a couple of weeks ago and solved it like this. @samuk190 I also forgot to add that you probably are gonna have to wrap those functions inside a useCallback hook so you can add them in the dependencies array of the useEffect.

Thanks for sharing your thoughts and ideas  @MuriloCampos . 

If someone here thinks is necessary to use useCallback  in their own app, please check this article before doing so https://kentcdodds.com/blog/usememo-and-usecallback, since every React app is different and using useCallback, may not be necessary.

**Comment by Saad-Bashar** [comment]
How about with the latest OneSignal function? I am trying to do this but the event listener is not getting fired. What am I missing, please? 

```
  React.useEffect(() => {
      OneSignal.setNotificationOpenedHandler(notification => {
        console.log("OneSignal: notification opened:", notification);
      });

      return OneSignal.clearHandlers()
    }, [])
```

**Comment by rgomezp** [comment]
Closing due to inactivity.

**Comment by xDemon200** [comment]
## Create custom hook called useOnesignal.ts
### for this hook use the latest released version **^4.5.0**

![image](https://user-images.githubusercontent.com/24462068/224812020-90bec31f-a984-4929-ada4-7d4da83c817a.png)

### Usage
![image](https://user-images.githubusercontent.com/24462068/224812204-c95bd3d6-3972-4369-8d13-ffe6613b35f2.png)
